File: src/arduinobot_test/src/simple_action_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include "arduinobot_test/FibonacciAction.h"

class FibonacciActionServer
{
    public:
        ros::NodeHandle nh;
        actionlib::SimpleActionServer<arduinobot_test::FibonacciAction> actionServer;
        std::string actionName;
        arduinobot_test::FibonacciFeedback feedback;
        arduinobot_test::FibonacciResult result;

        void executeCB(const arduinobot_test::FibonacciGoalConstPtr &goal)
        {
            ROS_INFO("Goal received %i", goal->goal);
            ros::Rate rate(1);
            bool success = true;
            feedback.sequence.clear();
            feedback.sequence.push_back(1);
            feedback.sequence.push_back(1);

            for (int i = 1; i < goal->goal - 1; i++)
            {
                if (actionServer.isPreemptRequested() || !ros::ok())
                {
                    ROS_INFO("%s is preempted", actionName.c_str());
                    actionServer.setPreempted();
                    success = false;
                    break;
                }

                feedback.sequence.push_back(feedback.sequence[i] + feedback.sequence[i - 1]);
                actionServer.publishFeedback(feedback);
                rate.sleep();
            }

            if (success)
            {
                result.sequence = feedback.sequence;
                ROS_INFO("%s has succeeded", actionName.c_str());
                actionServer.setSucceeded(result);
            }
        }

        FibonacciActionServer(std::string name) : actionServer(nh, name, boost::bind(&FibonacciActionServer::executeCB, this, _1), false),
                                                actionName(name)
        {
            actionServer.start();
            ROS_INFO("Simple action server started");
        }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "fibonacci");
    FibonacciActionServer fibonacci("fibonacci");

    ros::spin();
    return 0;
}

File: src/arduinobot_test/scripts/simple_action_server.py
#!/usr/env/bin python3
import rospy
import actionlib
from arduinobot_test.msg import FibonacciAction, FibonacciResult, FibonacciFeedback

class FibonacciActionServer(object):
    feedback_ = FibonacciFeedback()
    result_ = FibonacciResult()

    def __init__(self, name):
        self.action_name = name
        self.as_ = actionlib.SimpleActionServer(self.action_name, FibonacciAction, execute_cb = self.execute_cb, auto_start = False)
        self.as_.start()
        rospy.loginfo('Simple Action Server Started')

    
    def execute_cb(self, goal):
        rospy.loginfo('Goal Received %s' % goal.goal)
        r = rospy.Rate(1)
        success = True
        self.feedback_.sequence = []
        self.feedback_.sequence.append(1)
        self.feedback_.sequence.append(1)

        for i in range(1, goal.goal - 1):
            if self.as_.is_preempt_requested():
                rospy.loginfo('%s Preempted' % self.action_name)
                self.as_.set_preempted()
                success = False
                break

            self.feedback_.sequence.append(self.feedback_.sequence[i] + self.feedback_.sequence[i -1])
            self.as_.publish_feedback(self.feedback_)
            r.sleep()

        if success:
            self.result_.sequence = self.feedback_.sequence
            rospy.loginfo('%s Succeeded' % self.action_name)
            self.as_.set_succeeded(self.result_)


if __name__ == '__main__':
    rospy.init_node('fibonacci')
    
    server = FibonacciActionServer('fibonacci')
    rospy.spin()